### PR TITLE
SAAS-3780: Added Enable Access Logging For App Mesh Virtual Gateways Plugin

### PR DESCRIPTION
**Pull request by sadeed12345**

### CODE REVIEW

File: plugins/aws/appmesh/enableAccessLogging.js [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+                if (!mesh.arn) continue;
+
+                let resource = mesh.arn;
+
+                var listVirtualGateways = helpers.addSource(cache, source,
+                    ['appmesh', 'listVirtualGateways', region, mesh.meshName]);
+
+                if (!listVirtualGateways || listVirtualGateways.err || !listVirtualGateways.data) {
+                    helpers.addResult(results, 3,
+                        'Unable to query for AppMesh virtual gateways: ' + helpers.addError(listVirtualGateways),
```
Comment (sadeed12345) [inline plugins/aws/appmesh/enableAccessLogging.js:48]:
```suggestion
                        'Unable to query for App Mesh virtual gateways: ' + helpers.addError(listVirtualGateways),
```

File: helpers/aws/regions.js [hunk 2]
```diff
@@ -185,5 +185,9 @@ module.exports = {
         'eu-west-2', 'ap-northeast-1', 'ap-northeast-2', 'ap-southeast-1', 'ap-southeast-2', 'af-south-1'],
     fsx: regions,
     wisdom: ['us-east-1', 'us-west-2', 'eu-west-2', 'eu-central-1', 'ap-northeast-1', 'ap-southeast-2'],
-    voiceid: ['us-east-1', 'us-west-2', 'eu-west-2', 'eu-central-1', 'ap-southeast-1', 'ap-northeast-1', 'ap-southeast-2']
+    voiceid: ['us-east-1', 'us-west-2', 'eu-west-2', 'eu-central-1', 'ap-southeast-1', 'ap-northeast-1', 'ap-southeast-2'],
+    appmesh:  ['us-east-1', 'us-east-2', 'us-west-1', 'us-west-2', 'ca-central-1', 'eu-central-1',
+        'eu-west-1', 'eu-west-2', 'eu-west-3', 'eu-north-1', 'ap-northeast-1', 'ap-northeast-2',
+        'ap-southeast-1', 'ap-southeast-2', 'ap-south-1', 'sa-east-1', 'ap-east-1', 'me-south-1'
+    ],
```
Comment (AkhtarAmir) [inline helpers/aws/regions.js:192]:
```suggestion
        'ap-southeast-1', 'ap-southeast-2', 'ap-south-1', 'sa-east-1', 'ap-east-1', 'me-south-1']
```

File: plugins/aws/appmesh/enableAccessLogging.js [hunk 3]
```diff
@@ -0,0 +1,99 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'Enable Access Logging for App Mesh Virtual Gateways',
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/enableAccessLogging.js:5]:
```suggestion
    title: 'App Mesh Virtual Gateway Access Logging',
```

File: plugins/aws/appmesh/enableAccessLogging.js [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    category: 'App Mesh',
+    domain: 'Content Delivery',
+    description: 'Ensure that your Amazon App Mesh virtual gateways have access logging enabled and configured for all.',
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/enableAccessLogging.js:8]:
```suggestion
    description: 'Ensure that your Amazon App Mesh virtual gateways have access logging enabled.',
```

File: plugins/aws/appmesh/enableAccessLogging.js [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    more_info: 'The Amazon App Mesh virtual gateways Access Logging feature provide evidence for security audits and investigations, it also lets you keep an eye on application mesh user access and helps you meet compliance regulations. ',
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/enableAccessLogging.js:9]:
```suggestion
    more_info: 'The Amazon App Mesh virtual gateways Access Logging feature provide evidence for security audits and investigations as well as let\'s you keep an eye on application mesh user access and helps you meet compliance regulations. ',
```

File: plugins/aws/appmesh/enableAccessLogging.js [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    link: 'https://docs.aws.amazon.com/app-mesh/latest/userguide/envoy-logs.html',
+    recommended_action: 'Enable the feature, configure the file path to write access logs, within the virtual gateway configuration settings.',
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/enableAccessLogging.js:11]:
```suggestion
    recommended_action: 'Modify App Mesh virtual gateways to configure access logging feature',
```

File: plugins/aws/appmesh/enableAccessLogging.js [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    apis: ['AppMesh:listMeshes', 'AppMesh:listVirtualGateways', 'AppMesh:describeVirtualGateway'],
+    
+
+    run: function(cache, settings, callback) {
+        var results = [];
+        var source = {};
+        var region = helpers.regions(settings);
+
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/enableAccessLogging.js:19]:
```suggestion
    run: function(cache, settings, callback) {
        var results = [];
        var source = {};
        var region = helpers.regions(settings);
```

File: plugins/aws/appmesh/enableAccessLogging.js [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+                }
+
+                if (!listVirtualGateways.data.virtualGateways || !listVirtualGateways.data.virtualGateways.length) {
+                    helpers.addResult(results, 0,
+                        'No AppMesh virtual gateways found', region, resource);
+                    continue;
+                }
+
+                for (let gateway of listVirtualGateways.data.virtualGateways) {
+                    if (!gateway.arn) continue;
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/enableAccessLogging.js:60]:
```suggestion
                    if (!gateway.arn || gateway.virtualGatewayName) continue;
```

File: plugins/aws/appmesh/enableAccessLogging.js [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+                    let resource = gateway.arn;
+
+                    var describeVirtualGateway = helpers.addSource(cache, source,
+                        ['appmesh', 'describeVirtualGateway', region, gateway.virtualGatewayName]);
+
+                    if (!describeVirtualGateway ||
+                        describeVirtualGateway.err ||
+                        !describeVirtualGateway.data) {
+                        helpers.addResult(results, 3,
+                            'Unable to query AppMesh virtual gateway description: ' + helpers.addError(describeVirtualGateway), region, resource);
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/enableAccessLogging.js:71]:
```suggestion
                            'Unable to describe AppMesh virtual gateway: ' + helpers.addError(describeVirtualGateway), region, resource);
```

File: plugins/aws/appmesh/enableAccessLogging.js [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+                        !describeVirtualGateway.data.virtualGateway.spec.logging.accessLog.file.path) {
+                        helpers.addResult(results, 2,
+                            'access logging is not enabled for Amazon App Mesh virtual gateways',
+                            region, resource);
+                        continue;         
+                    } else {
+                        helpers.addResult(results, 0,
+                            'access logging is enabled and configured for Amazon App Mesh virtual gateways',
+                            region, resource);
+                        continue;
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/enableAccessLogging.js:89]:
```suggestion
                    if (describeVirtualGateway.data.virtualGateway &&
                        describeVirtualGateway.data.virtualGateway.spec &&
                        describeVirtualGateway.data.virtualGateway.spec.logging &&
                        describeVirtualGateway.data.virtualGateway.spec.logging.accessLog &&
                        describeVirtualGateway.data.virtualGateway.spec.logging.accessLog.file &&
                        describeVirtualGateway.data.virtualGateway.spec.logging.accessLog.file.path &&
                        describeVirtualGateway.data.virtualGateway.spec.logging.accessLog.file.path.length) {
                        helpers.addResult(results, 0,
                            'App Mesh virtual gateway has access logging enabled',
                            region, resource);
                        continue;         
                    } else {
                        helpers.addResult(results, 2,
                            'App Mesh virtual gateway does not have access logging enabled',
                            region, resource);
                        continue;
```

File: plugins/aws/appmesh/enableAccessLogging.spec.js [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+
+describe('enableAccessLogging', function () {
+    describe('run', function () {
+
+        it('should PASS if access logging is enabled and configured for Amazon App Mesh virtual gateways', function (done) {
+            const cache = createCache([listMeshes[0]], [listVirtualGateways[1]], describeVirtualGateway[1]);
+            enableAccessLogging.run(cache, { }, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(0);
+                expect(results[0].message).to.include('access logging is enabled and configured for Amazon App Mesh virtual gateways');
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/enableAccessLogging.spec.js:177]:
```suggestion
                expect(results[0].message).to.include('App Mesh virtual gateway has access logging enabled');
```

File: plugins/aws/appmesh/enableAccessLogging.spec.js [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+                done();
+            });
+        });
+
+        it('should FAIL if access logging is not enabled for Amazon App Mesh virtual gateways', function (done) {
+            const cache = createCache([listMeshes[0]], [listVirtualGateways[0]], describeVirtualGateway[0]);
+            enableAccessLogging.run(cache, { }, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(2);
+                expect(results[0].message).to.include('access logging is not enabled for Amazon App Mesh virtual gateways');
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/enableAccessLogging.spec.js:188]:
```suggestion
                expect(results[0].message).to.include('App Mesh virtual gateway does not have access logging enabled');
```

File: plugins/aws/appmesh/enableAccessLogging.spec.js [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+        it('should PASS if No App Meshes found', function (done) {
+            const cache = createCache([]);
+            enableAccessLogging.run(cache, { }, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].message).to.include('No App Meshes found');
+                expect(results[0].status).to.equal(0);
+                done();
+            });
+        });
+
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/enableAccessLogging.spec.js:203]: In below tests, order doesn't seem right. Also add check for message

File: plugins/aws/appmesh/enableAccessLogging.js [hunk 14]
```diff
@@ -0,0 +1,98 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/enableAccessLogging.js:3]: Change plugin file name to 'appmeshVGAccessLogging'

**Review by AkhtarAmir — CHANGES_REQUESTED** [review]

File: plugins/aws/appmesh/appmeshVGAccessLogging.js [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+
+        async.each(region.appmesh, function(region, rcb){
+            var listMeshes = helpers.addSource(cache, source,
+                ['appmesh', 'listMeshes', region]);
+            
+            if (!listMeshes) return rcb();
+
+            if (listMeshes.err || !listMeshes.data) {
+                helpers.addResult(results, 3,
+                    'Unable to query for App Meshes: ' + helpers.addError(listMeshes),region);
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/appmeshVGAccessLogging.js:27]:
```suggestion
                    'Unable to query for App Mesh meshes: ' + helpers.addError(listMeshes),region);
```

File: plugins/aws/appmesh/appmeshVGAccessLogging.js [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+                return rcb();
+            }
+
+            if (!listMeshes.data.length) {
+                helpers.addResult(results, 0, 'No App Meshes found', region);
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/appmeshVGAccessLogging.js:32]:
```suggestion
                helpers.addResult(results, 0, 'No App Mesh meshes found', region);
```

File: plugins/aws/appmesh/appmeshVGAccessLogging.js [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+                    let resource = gateway.arn;
+
+                    var describeVirtualGateway = helpers.addSource(cache, source,
+                        ['appmesh', 'describeVirtualGateway', region, gateway.virtualGatewayName]);
+
+                    if (!describeVirtualGateway ||
+                        describeVirtualGateway.err ||
+                        !describeVirtualGateway.data) {
+                        helpers.addResult(results, 3,
+                            'Unable to describe AppMesh virtual gateway: ' + helpers.addError(describeVirtualGateway), region, resource);
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/appmeshVGAccessLogging.js:69]:
```suggestion
                            'Unable to describe App Mesh virtual gateway: ' + helpers.addError(describeVirtualGateway), region, resource);
```

File: plugins/aws/appmesh/appmeshVGAccessLogging.js [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+                if (!listVirtualGateways || listVirtualGateways.err || !listVirtualGateways.data) {
+                    helpers.addResult(results, 3,
+                        'Unable to query for App Mesh virtual gateways: ' + helpers.addError(listVirtualGateways),
+                        region, resource);
+                    continue;
+                }
+
+                if (!listVirtualGateways.data.virtualGateways ) {
+                    helpers.addResult(results, 0,
+                        'No AppMesh virtual gateways found', region, resource);
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/appmeshVGAccessLogging.js:53]:
```suggestion
                        'No App Mesh virtual gateways found', region, resource);
```

File: plugins/aws/appmesh/appmeshVGAccessLogging.js [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+                    ['appmesh', 'listVirtualGateways', region, mesh.meshName]);
+
+                if (!listVirtualGateways || listVirtualGateways.err || !listVirtualGateways.data) {
+                    helpers.addResult(results, 3,
+                        'Unable to query for App Mesh virtual gateways: ' + helpers.addError(listVirtualGateways),
+                        region, resource);
+                    continue;
+                }
+
+                if (!listVirtualGateways.data.virtualGateways ) {
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/appmeshVGAccessLogging.js:51]:
```suggestion
                if (!listVirtualGateways.data.virtualGateways) {
```

File: plugins/aws/appmesh/appmeshVGAccessLogging.spec.js [hunk 20]
```diff
@@ -0,0 +1,232 @@
...
+                expect(results[0].region).to.equal('us-east-1');
+                done();
+            });
+        });
+
+        it('should PASS if No App Meshes found', function (done) {
+            const cache = createCache([]);
+            appmeshVGAccessLogging.run(cache, { }, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].message).to.include('No App Meshes found');
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/appmeshVGAccessLogging.spec.js:198]:
```suggestion
                expect(results[0].message).to.include('No App Mesh meshes found');
```

File: plugins/aws/appmesh/appmeshVGAccessLogging.spec.js [hunk 21]
```diff
@@ -0,0 +1,232 @@
...
+                expect(results[0].status).to.equal(0);
+                done();
+            });
+        });
+
+        it('should UNKNOWN if unable to query for App Meshes', function (done) {
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/appmeshVGAccessLogging.spec.js:204]:
```suggestion
        it('should UNKNOWN if unable to query for App Mesh meshes', function (done) {
```

File: plugins/aws/appmesh/appmeshVGAccessLogging.spec.js [hunk 22]
```diff
@@ -0,0 +1,232 @@
...
+            const cache = createCache(null, null, null, { message: 'Unable to query for App Meshes'});
+            appmeshVGAccessLogging.run(cache, { }, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(3);
+                expect(results[0].message).to.include('Unable to query for App Meshes');
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/appmeshVGAccessLogging.spec.js:209]:
```suggestion
                expect(results[0].message).to.include('Unable to query for App Mesh meshes');
```

File: plugins/aws/appmesh/appmeshVGAccessLogging.spec.js [hunk 23]
```diff
@@ -0,0 +1,232 @@
...
+                done();
+            });
+        });
+
+        it('should UNKNOWN if unable to query for AppMesh virtual gateways', function (done) {
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/appmeshVGAccessLogging.spec.js:214]:
```suggestion
        it('should UNKNOWN if unable to query for App Mesh virtual gateways', function (done) {
```

**Review by AkhtarAmir — APPROVED** [review]

File: plugins/aws/appmesh/appmeshVGAccessLogging.js [hunk 24]
```diff
@@ -0,0 +1,96 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'App Mesh VG Access Logging',
+    category: 'App Mesh',
+    domain: 'Content Delivery',
+    description: 'Ensure that your Amazon App Mesh virtual gateways have access logging enabled.',
+    more_info: 'The Amazon App Mesh virtual gateways Access Logging feature provide evidence for security audits and investigations as well as let\'s you keep an eye on application mesh user access and helps you meet compliance regulations. ',
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/appmeshVGAccessLogging.js:9]:
```suggestion
    more_info: 'Amazon App Mesh virtual gateways Access Logging feature provide evidence for security audits and investigations as well as let\'s you keep an eye on application mesh user access and helps you meet compliance regulations. ',
```

File: plugins/aws/appmesh/appmeshVGAccessLogging.js [hunk 25]
```diff
@@ -0,0 +1,96 @@
...
+    link: 'https://docs.aws.amazon.com/app-mesh/latest/userguide/envoy-logs.html',
+    recommended_action: 'Modify App Mesh virtual gateways to configure access logging feature',
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/appmeshVGAccessLogging.js:11]:
```suggestion
    recommended_action: 'To enable access logging, modify virtual gateway configuration settings and configure the file path to write access logs to.',
```

File: plugins/aws/appmesh/appmeshVGAccessLogging.js [hunk 26]
```diff
@@ -0,0 +1,96 @@
...
+                return rcb();
+            }
+
+            if (!listMeshes.data.length) {
+                helpers.addResult(results, 0, 'No App Mesh meshes found', region);
+                return rcb();
+            }
+
+            for (let mesh of listMeshes.data){
+                if (!mesh.arn) continue;
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/appmeshVGAccessLogging.js:37]:
```suggestion
                if (!mesh.arn || !mesh.meshName) continue;
```

File: plugins/aws/appmesh/appmeshVGAccessLogging.js [hunk 27]
```diff
@@ -0,0 +1,96 @@
...
+                }
+
+                if (!listVirtualGateways.data.virtualGateways) {
+                    helpers.addResult(results, 0,
+                        'No App Mesh virtual gateways found', region, resource);
+                    continue;
+                }
+
+                for (let gateway of listVirtualGateways.data.virtualGateways) {
+                    if (!gateway.arn) continue;
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/appmeshVGAccessLogging.js:58]:
```suggestion
                    if (!gateway.arn || gateway.virtualGatewayName) continue;
```

File: plugins/aws/appmesh/appmeshVGAccessLogging.spec.js [hunk 28]
```diff
@@ -0,0 +1,232 @@
+const expect = require('chai').expect;
+var appmeshVGAccessLogging = require('./appmeshVGAccessLogging');
+
+
```
Comment (AkhtarAmir) [inline plugins/aws/appmesh/appmeshVGAccessLogging.spec.js:4]:
```suggestion
```

**Review by AkhtarAmir — APPROVED** [review]

**Review by AkhtarAmir — APPROVED** [review]

**Pull request merged**